src/reply_manager.rs

```diff
@@ -0,0 +1,228 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! A mixnode may receive the same request multiple times due to retransmission (see eg
+//! [`request_manager`](super::request_manager)). A [`ReplyManager`] can be used to cache replies,
+//! to avoid needing to execute requests more than once.
+
+use super::core::{MessageId, Mixnet, RequestMessage, SessionIndex, Surb, MESSAGE_ID_SIZE};
+use hashlink::{linked_hash_map::Entry, LinkedHashMap};
+use log::{debug, warn};
+use rand::RngCore;
+use std::time::{Duration, Instant};
+
+/// Reply manager configuration.
+#[derive(Clone, Debug)]
+pub struct Config {
+	/// The target for log messages.
+	pub log_target: &'static str,
+	/// Maximum number of requests to remember. When this limit is reached, old requests will be
+	/// automatically discarded to make space for new ones.
+	pub capacity: usize,
+	/// Maximum number of copies of a reply message to post in response to a single request
+	/// message. Note that the number of copies is also limited by the number of SURBs provided in
+	/// the request message.
+	pub max_posts: usize,
+	/// After replying to a request, ignore repeats of the request for this length of time. This
+	/// should ideally be set such that extra copies of a request message posted at the same time
+	/// as the first received one get ignored, but retries posted after a timeout do not.
+	pub cooldown: Duration,
```

Comment:
`max_posts` is somewhat configurable on a per-reply basis as you can limit the number of SURBs you include in the request message. `cooldown` could be as well, though I'm not sure you'd want to do that -- I think you can choose a value that catches 99% of redundant replies without discarding needed replies, without needing to know anything particular about the request. I would generally be careful about including unnecessary information in requests because any extra information that is included could compromise anonymity.

Reply:
Sorry I thought you meant configured by the node sending the request. I see now that you meant by the code sending the reply. I agree these things could be controlled more finely, but I don't currently see a need for this?

Reply:
Sure probably better to have it when needed.